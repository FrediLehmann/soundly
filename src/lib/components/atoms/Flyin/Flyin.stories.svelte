<script>
  import { Meta, Story, Template } from '@storybook/addon-svelte-csf';

  import { expect } from '@storybook/jest';
  import {
    within,
    userEvent,
    waitForElementToBeRemoved
  } from '@storybook/testing-library';

  import Flyin from './Flyin.svelte';

  let flyin;
</script>

<Meta
  title="Atoms/Flyin"
  component={Flyin}
  args={{ message: 'Information for the user.' }}
  argTypes={{
    message: {
      control: { type: 'text' },
      table: {
        type: {
          summary: 'Text to show in the flyin',
          detail: 'What information is shown to the user.'
        }
      }
    },
    style: {
      control: { type: 'select' },
      options: ['success', 'error', 'info'],
      table: {
        type: {
          summary: 'Type of flyin message',
          detail:
            'The flyin is used to show different information to the user, based on events. The style is of this message is indication what kind of event happended.'
        }
      }
    },
    showFor: {
      control: { type: 'number' },
      table: {
        type: {
          summary: 'Time for which the flyin should be shown'
        }
      }
    }
  }} />

<Template let:args>
  <Flyin bind:this={flyin} />
  <button
    class="btn"
    on:click={() => flyin.show(args.message, args.style, 5000)}>
    Show flyin
  </button>
</Template>

<Story
  name="Success"
  args={{ style: 'success' }}
  play={async ({ canvasElement }) => {
    const canvas = within(canvasElement);

    userEvent.click(canvas.getByRole('button'));
    expect(canvas.getByText('Information for the user.')).toBeTruthy();

    userEvent.click(canvas.getByText('Close'));
    expect(
      waitForElementToBeRemoved(canvas.getByText('Information for the user.'))
    ).resolves;

    userEvent.click(canvas.getByText('Show flyin'));
    expect(canvas.getByText('Information for the user.')).toBeTruthy();
  }} />

<Story
  name="Error"
  args={{ style: 'error' }}
  play={async ({ canvasElement }) => {
    const canvas = within(canvasElement);

    userEvent.click(canvas.getByRole('button'));
    expect(canvas.getByText('Information for the user.')).toBeTruthy();

    userEvent.click(canvas.getByText('Close'));
    expect(
      waitForElementToBeRemoved(canvas.getByText('Information for the user.'))
    ).resolves;

    userEvent.click(canvas.getByText('Show flyin'));
    expect(canvas.getByText('Information for the user.')).toBeTruthy();
  }} />

<Story
  name="Info"
  args={{ style: 'info' }}
  play={async ({ canvasElement }) => {
    const canvas = within(canvasElement);

    userEvent.click(canvas.getByRole('button'));
    expect(canvas.getByText('Information for the user.')).toBeTruthy();

    userEvent.click(canvas.getByText('Close'));
    expect(
      waitForElementToBeRemoved(canvas.getByText('Information for the user.'))
    ).resolves;

    userEvent.click(canvas.getByText('Show flyin'));
    expect(canvas.getByText('Information for the user.')).toBeTruthy();
  }} />

<style>
  .btn {
    margin-top: 100px;
    border: 1px solid black;
    background-color: transparent;
    color: black;
  }
</style>
