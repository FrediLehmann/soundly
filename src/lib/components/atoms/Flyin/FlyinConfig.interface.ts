export interface FlyinConfig {
    message: string
}