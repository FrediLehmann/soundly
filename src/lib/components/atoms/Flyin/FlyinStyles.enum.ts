export enum FlyinStyles {
  info = 'info',
  error = 'error',
  success = 'success'
}
