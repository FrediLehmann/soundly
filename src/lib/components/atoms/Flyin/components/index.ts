import Icon from './Icon.svelte';
import CloseButton from './CloseButton.svelte';

export { CloseButton, Icon };
