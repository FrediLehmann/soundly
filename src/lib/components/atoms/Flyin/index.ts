import Flyin from './Flyin.svelte';
import { FlyinStyles } from './FlyinStyles.enum';
import type { FlyinConfig } from './FlyinConfig.interface';

export { Flyin, FlyinStyles };
export type { FlyinConfig };
