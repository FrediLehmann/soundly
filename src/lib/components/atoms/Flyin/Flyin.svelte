<script lang="ts" context="module">
  import type { FlyinConfig } from './FlyinConfig.interface';
  import type { FlyinStyles } from './FlyinStyles.enum';

  import { fly } from 'svelte/transition';
  import { Icon, CloseButton } from './components';
</script>

<script lang="ts">
  let isVisible = false;
  let flyinMessage: string;
  let flyinStyle: FlyinStyles;
  let flyinTimeout: number;

  export const show = ({
    message,
    timeout = 5000,
    style
  }: FlyinConfig): void => {
    flyinMessage = message;
    flyinStyle = style;
    isVisible = true;

    flyinTimeout = window.setTimeout(() => (isVisible = false), timeout);
  };

  const onClose = (): void => {
    window.clearTimeout(flyinTimeout);
    isVisible = false;
  };
</script>

{#if isVisible}
  <div
    class="flyinContainer"
    data-type={flyinStyle}
    transition:fly={{ y: -100, duration: 300 }}>
    <div class="flyinContent">
      <Icon {flyinStyle} />
      <span>{flyinMessage}</span>
      <CloseButton on:click={onClose} />
    </div>
  </div>
{/if}
