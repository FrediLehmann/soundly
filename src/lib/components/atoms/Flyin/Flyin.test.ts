/**
 * @jest-environment jsdom
 */

import '@testing-library/jest-dom';
import { render } from '@testing-library/svelte';

import Flyin from './Flyin.svelte';

describe('<Flyin /> component', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.runOnlyPendingTimers();
    jest.useRealTimers();
  });

  test('component not in DOM on default', () => {
    const { container } = render(Flyin);
    expect(container.innerHTML).toEqual('<div></div>');
  });

  //// Tests currently not updating state correctly

  //   test('component visible on show', async () => {
  //     const { container, component, getByText, getByRole } = render(Flyin);
  //     await act(() =>
  //       component.show({
  //         message: 'Test Show',
  //         timeout: 20,
  //         style: FlyinStyles.info
  //       })
  //     );
  //     expect(getByText('Test Show')).toBeInTheDocument();

  //     await fireEvent.click(getByRole('button'));
  //     expect(container.innerHTML).toEqual('<div></div>');
  //   });

  //   test('hdes after timeout', async () => {
  //     const { container, component, getByText } = render(Flyin);
  //     await act(() =>
  //       component.show({
  //         message: 'Test Show',
  //         timeout: 1,
  //         style: FlyinStyles.info
  //       })
  //     );

  //     expect(getByText('Test Show')).toBeInTheDocument();

  //     await jest.advanceTimersByTime(2);
  //     expect(container.innerHTML).toEqual('<div></div>');
  //   });
});
