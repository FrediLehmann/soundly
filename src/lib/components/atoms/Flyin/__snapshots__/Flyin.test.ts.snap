// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Flyin /> component component not in DOM on default 1`] = `
Component {
  "$$": Object {
    "after_update": Array [],
    "before_update": Array [],
    "bound": Object {},
    "callbacks": Object {},
    "context": Map {},
    "ctx": Array [
      Flyin {
        "$$": Object {
          "after_update": Array [],
          "before_update": Array [],
          "bound": Object {},
          "callbacks": Object {},
          "context": Map {},
          "ctx": Array [
            true,
            "Test",
            "info",
            [Function],
            [Function],
          ],
          "dirty": Array [
            7,
          ],
          "fragment": Object {
            "c": [Function],
            "d": [Function],
            "i": [Function],
            "l": [Function],
            "m": [Function],
            "o": [Function],
            "p": [Function],
          },
          "not_equal": [Function],
          "on_destroy": Array [],
          "on_disconnect": Array [],
          "on_mount": Array [],
          "props": Object {
            "show": 4,
          },
          "root": <div>
            
             
            <button>
              Open
            </button>
          </div>,
          "skip_bound": false,
          "update": [Function],
        },
        "$capture_state": [Function],
        "$inject_state": [Function],
      },
      [Function],
      [Function],
    ],
    "dirty": Array [
      -1,
    ],
    "fragment": Object {
      "c": [Function],
      "d": [Function],
      "i": [Function],
      "l": [Function],
      "m": [Function],
      "o": [Function],
      "p": [Function],
    },
    "not_equal": [Function],
    "on_destroy": Array [
      [Function],
    ],
    "on_disconnect": Array [],
    "on_mount": Array [],
    "props": Object {},
    "root": <div>
      
       
      <button>
        Open
      </button>
    </div>,
    "skip_bound": false,
    "update": [Function],
  },
  "$capture_state": [Function],
  "$inject_state": [Function],
}
`;
